Add support for insert(elem)

zaporter opened the pull request:
The ability to insert into this structure can be very important if you have a read-heavy but right-necessary workload where rebuilding the whole data structure for every write is infeasible.

I have tried to implement this function so that it is easy to understand and maintain. It might not be as fast as it could be (that is outside my area of expertise). 

Please let me know if you have intentionally avoided adding this function for whatever reason or if you think that I should be using a different data structure. 

I also refactored some of the tests. I did this just in case you want to add other methods in the future, it seemed like the right thing to do. Let me know if you have any opinions on that. 


Thank you for your cool library :heart:

Commit:
Add support for insert(elem)

Why: The ability to insert into this structure can be very important if
you have a read-heavy but right-necessary workload where rebuilding
the whole data structure would take too long

sstadick commented on src/lib.rs:
```diff
@@ -955,6 +1085,43 @@ mod tests {
         assert_eq!(expected, lapper.iter().collect::<Vec<&Iv>>());
         assert_eq!(lapper.intervals.len(), lapper.starts.len())
 
+    }
```
```suggestion
    }
    
```

sstadick commented on src/lib.rs:
```diff
@@ -955,6 +1085,43 @@ mod tests {
+        lapper.merge_overlaps();
+        assert_eq!(expected, lapper.iter().collect::<Vec<&Iv>>());
+        assert_eq!(lapper.intervals.len(), lapper.starts.len())
+
+    }
```
```suggestion
    }

```

sstadick commented on src/lib.rs:
```diff
@@ -955,6 +1085,43 @@ mod tests {
+        lapper.merge_overlaps();
+        assert_eq!(expected, lapper.iter().collect::<Vec<&Iv>>());
+        assert_eq!(lapper.intervals.len(), lapper.starts.len())
+
```
```suggestion
```

sstadick commented on src/lib.rs:
```diff
@@ -996,6 +1163,18 @@ mod tests {
         lapper.set_cov();
         assert_eq!(lapper.cov(), 50);
     }
+    #[test]
```
```suggestion

    #[test]
```

sstadick commented on src/lib.rs:
```diff
@@ -939,6 +1048,27 @@ mod tests {
         );
         assert_eq!(lapper.count(8, 20), 2);
     }
+    #[test]
```
```suggestion

    #[test]
```

sstadick commented on src/lib.rs:
```diff
@@ -872,6 +967,20 @@ mod tests {
         assert_eq!(Some(&expected), lapper.seek(15, 25, &mut cursor).next());
         assert_eq!(lapper.find(15, 25).count(), lapper.count(15, 25));
     }
+    // Test that a query that overlaps the start of an interval returns that interval
```
```suggestion

    // Test that a query that overlaps the start of an interval returns that interval
```

sstadick commented on src/lib.rs:
```diff
@@ -857,6 +943,15 @@ mod tests {
         assert_eq!(None, lapper.seek(30, 35, &mut cursor).next());
         assert_eq!(lapper.find(30, 35).count(), lapper.count(30, 35));
     }
+    // Test that a query start that hits an interval end returns no interval
```
```suggestion

    // Test that a query start that hits an interval end returns no interval
```

sstadick commented on src/lib.rs:
```diff
@@ -847,6 +924,15 @@ mod tests {
         assert_eq!(None, lapper.seek(15, 20, &mut cursor).next());
         assert_eq!(lapper.find(15, 20).count(), lapper.count(15, 20));
     }
+    // Test that a query stop that hits an interval start returns no interval
```
```suggestion

    // Test that a query stop that hits an interval start returns no interval
```

sstadick commented on src/lib.rs:
```diff
@@ -826,8 +893,18 @@ mod tests {
+
+    fn setup_badlapper() -> Lapper<usize, u32> {
+        Lapper::new(get_data_badlapper())
+    }
```
```suggestion
    }

```

sstadick commented on src/lib.rs:
```diff
@@ -792,30 +840,49 @@ mod tests {
+        }
+        lapper
+
+    }
```
```suggestion
    }

```

sstadick commented on src/lib.rs:
```diff
@@ -792,30 +840,49 @@ mod tests {
+    fn setup_overlapping() -> Lapper<usize, u32> {
+        Lapper::new(get_data_overlapping())
+    }
+    fn setup_overlapping_insert() -> Lapper<usize, u32> {
```
```suggestion

    fn setup_overlapping_insert() -> Lapper<usize, u32> {
```

sstadick commented on src/lib.rs:
```diff
@@ -792,30 +840,49 @@ mod tests {
+    fn setup_nonoverlapping() -> Lapper<usize, u32> {
+        Lapper::new(get_data_nonoverlapping())
+    }
+    fn setup_nonoverlapping_insert() -> Lapper<usize, u32> {
```
```suggestion

    fn setup_nonoverlapping_insert() -> Lapper<usize, u32> {
```

sstadick commented on src/lib.rs:
```diff
@@ -792,30 +840,49 @@ mod tests {
     }
-    fn setup_overlapping() -> Lapper<usize, u32> {
-        let data: Vec<Iv> = (0..100)
+    fn setup_nonoverlapping() -> Lapper<usize, u32> {
```
```suggestion

    fn setup_nonoverlapping() -> Lapper<usize, u32> {
```

sstadick commented on src/lib.rs:
```diff
@@ -792,30 +840,49 @@ mod tests {
+        }
+        lapper
+    }
+    fn get_data_overlapping() -> Vec<Iv> {
```
```suggestion

    fn get_data_overlapping() -> Vec<Iv> {
```

sstadick commented on src/lib.rs:
```diff
@@ -792,30 +840,49 @@ mod tests {
     type Iv = Interval<usize, u32>;
-    fn setup_nonoverlapping() -> Lapper<usize, u32> {
-        let data: Vec<Iv> = (0..100)
+    fn get_data_nonoverlapping() -> Vec<Iv> {
```
```suggestion

    fn get_data_nonoverlapping() -> Vec<Iv> {
```

sstadick commented on src/lib.rs:
```diff
@@ -373,16 +410,27 @@ where
+    {
+        Self::bsearch_seq_ref(&key, elems)
+    }
+    #[inline]
```
```suggestion

    #[inline]
```

sstadick reviewed (approved):
@zaporter Thanks for the contribution! This is excellent, I had just a few nit-picky changes that I've proposed (apologies if you were following formatting that I already had in place, I've evolved!)

Assuming this passes all tests I'll merge it once the changes are in and cut a release. Thank you!

sstadick commented:
@zaporter - lookslike it's just failing the cargofmt lint. Could you run that on your branch and push to here? I think we should be in a good spot after that.

sstadick commented:
I only never added an insert because I didn't have that use case and I knew it would be slooooow. But I have nothing against having it available with the appropriate caveats in place in the docs 👍

Commit:
Updated code to follow formatting standards

Mostly fixed errors with newlines. This change also edited the doc
comment for insert() to be more clear.

zaporter commented:
@sstadick Thanks for the formatting advice! I forgot to run a format check before submitting this :facepalm: . 

Let me know if you have any further comments, but I think I addressed everything.
 
Re: not implemented: Yeah, an insert is slow, but hopefully this will still be faster than other data structures for a read heavy workload. I need to do some benchmarking... If I get any interesting numbers I might submit them just so other people know. 

Again, thanks for making this!

Commit:
Updated tests to properly reflect desired outcome

Why: Previously the tests were testing whether or not the new
insert() mechanism had the same outcome when doing other methods.
That is actually not the desired API of insert. Insert should
be the exact same as if the data was included in the first place
and so it should check for that.

zaporter commented:
I woke up this morning and realized that my tests weren't testing what they should be. Just rewrote the tests. This should be a no-brainer switch. 

All of the tests pass and it guarantees that creating from ::new() or creating from ::insert() create the same Lapper. 

Ready to merge after your approval.

sstadick commented:
@zaporter This all looks great, thank you again! I'll make the release tomorrow 👍 and I haven't done it by tomorrow feel free to ping me.

sstadick commented:
@zaporter released, see v1.1.0. Thanks again!
